poetry/utils/env.py

```diff
@@ -758,20 +759,25 @@ def pip_version(self):
 
     @property
     def site_packages(self):  # type: () -> Path
-        # It seems that PyPy3 virtual environments
-        # have their site-packages directory at the root
-        if self._path.joinpath("site-packages").exists():
-            return self._path.joinpath("site-packages")
-
-        if self._is_windows:
-            return self._path / "Lib" / "site-packages"
-
-        return (
-            self._path
-            / "lib"
-            / "python{}.{}".format(*self.version_info[:2])
-            / "site-packages"
-        )
+        if self._site_packages is None:
+            site_packages = []
+            dist_packages = []
+            for entry in self.sys_path:
+                entry = Path(entry)
+                if entry.name == "site-packages":
+                    site_packages.append(entry)
+                elif entry.name == "dist-packages":
+                    dist_packages.append(entry)
+
+            if not site_packages and not dist_packages:
+                raise RuntimeError("Unable to find the site-packages directory")
+
+            if site_packages:
+                self._site_packages = site_packages[0]
+            else:
+                self._site_packages = dist_packages[0]
```

Comment:
I don't know where the `site_packages` property is used, so it isn't obvious to me, why only the first site_package is stored and why `dist-packages` are never stored when site_packages where found.

Reply:
The `site_packages` property is used in the `InstalledRepository` class (see https://github.com/python-poetry/poetry/blob/master/poetry/repositories/installed_repository.py#L32).

We only need the "site-packages" directory where `pip` will install the dependencies which is, as far as I know, the `site-packages` in virtual environments and most OS and `dist-packages` for Debian-based distributions.

Reply:
Ah, ok I got it. I wonder if something with [sysconfig](https://docs.python.org/3/library/sysconfig.html) would be more reliable? But as far as I can see, we must be able to detect if we are inside a virtual environment and of we are on a posix or nt system, to get the correct path (`purelib` seems to be the one we are interested in), are we?

Reply:
Thanks for pointing me to `sysconfig`! Do you know if there are some systems where `purelib` and `platlib` are different? If there is none we could indeed use `purelib` and avoid the heuristics we are currently doing.

Reply:
I'm not sure. I've just discovered that module as well. There is something in [PEP427](https://www.python.org/dev/peps/pep-0427/#what-s-the-deal-with-purelib-vs-platlib) about the difference.

Reply:
In the end, it seems that all installed packages will reside in `purelib` and/or `platlib`, so to get populate the `InstalledRepository` we can just check both, I think. I'll test it out and see where it goes.

Reply:
After testing it out it seems that using `sysconfig` won't help us. While it works in some cases, for Debian-based systems, for instance, it won't return the directory we need. We could use `site.getsitepackages()` for non-virtual environments but we would need to find a way to replicate it for virtual environments which would lead to using the same heuristic this PR already uses.

I think we can leave the PR as-is for now since I tested it on Ubuntu, MacOS and Windows without any issue and it fixes the issues I mentioned above.